codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -183,28 +204,45 @@ public void incrementWindowSize(ChannelHandlerContext ctx, Http2Stream stream, i
             // Update the stream window and write any pending frames for the stream.
             AbstractState state = state(stream);
             state.incrementStreamWindow(delta);
-            state.writeBytes(state.writableWindow());
+
+            assert !isWriting;
+            isWriting = true;
+            try {
+                state.writeBytes(writableBytes(state.writableWindow()));
+            } finally {
+                isWriting = false;
+            }
         }
     }
 
     @Override
-    public void sendFlowControlled(ChannelHandlerContext ctx, Http2Stream stream, FlowControlled frame) {
-        checkNotNull(ctx, "ctx");
+    public void sendFlowControlled(Http2Stream stream, FlowControlled frame) {
         checkNotNull(frame, "frame");
-        if (this.ctx != null && this.ctx != ctx) {
-            throw new IllegalArgumentException("Writing data from multiple ChannelHandlerContexts is not supported");
-        }
-        // Save the context. We'll use this later when we write pending bytes.
-        this.ctx = ctx;
         final AbstractState state;
         try {
             state = state(stream);
             state.enqueueFrame(frame);
         } catch (Throwable t) {
-            frame.error(t);
+            frame.error(ctx, t);
             return;
         }
-        state.writeBytes(state.writableWindow());
+
+        assert !isWriting;
+        isWriting = true;
+        try {
+            state.writeBytes(writableBytes(state.writableWindow()));
+        } finally {
+            isWriting = false;
+        }
+    }
+
+    @Override
+    public void writabilityChanged() throws Http2Exception {
+        // Writability is expected to change while we are writing. We cannot allow this event to trigger reentering
+        // the allocation and write loop. Reentering the event loop will lead to over or illegal allocation.
+        if (!isWriting) {
```

Comment:
I think this is the only place where we are providing re-entrant protection.  The other instances are just `assert` statements. I should be able to change this to `ctx.channel().isWritable()` and get rid of the `isWriting` for this PR, but I'm concerned about @louiscryan's PR...if it flushes during the allocation/write cycle we can't rely on the `ctx.channel().isWritable()`.  I need to investigate this more and work out what order we plan to pull in the PRs.